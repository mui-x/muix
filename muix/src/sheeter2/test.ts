import { compile } from './compiler2'
import { TSheeterSource, TSheeterCompiled } from './types'

const merging: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    $before: {
        before: 'before',
        self: 'before',
        web: 'before',
        after: 'before',
    },
    self: 'self',
    web: 'self',
    after: 'self',
    $web: {
        web: 'web',
        after: 'web',
    },
    $after: {
        after: 'after',
    },
}

const whenUsed: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    $whenUsed: {
        root: {
            plain: 'plain',
        }
    },
    ':hover': {
        $whenUsed: {
            root: {
                hover: 'hover',
            }
        },
    }
}

const whenUsedRecursion: TSheeterSource.RulesTreeRoot<'root' | 'disabled'> = {
    rule: 'rule',
    $whenUsed: {
        root: {
            plainRoot: 'plainRoot',
            $whenUsed: {
                disabled: {
                    plainInner: 'plainInner',
                    ':hover': {
                        plainInnerHover: 'plainInnerHover',
                        ':active': {
                            plainInnerHoverActive: 'plainInnerHoverActive',
                        }
                    }
                }
            },
        }
    },
    ':hover': {
        hover: 'hover',
        $whenUsed: {
            root: {
                hoverRoot: 'hoverRoot',
                $whenUsed: {
                    disabled: {
                        hoverRootInner: 'hoverRootInner',
                    }
                },
                ':active': {
                    hoverRootActive: 'hoverRootActive',
                    $whenUsed: {
                        disabled: {
                            hoverRootActiveInner: 'hoverRootActiveInner',
                        }
                    },
                },
            }
        },
    }
}



export const run = () => {
    let res
    //res = JSON.stringify(compile(merging, 'root'), null, 2)
    //res = JSON.stringify(compile(whenUsed, 'root'), null, 2)
    res = JSON.stringify(compile(whenUsedRecursion, 'root'), null, 2)
    debugger
}

