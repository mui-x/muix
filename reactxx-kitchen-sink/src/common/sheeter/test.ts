
import * as Sheeter from 'reactxx-sheeter'

export const test = () => {

  const patchable = Sheeter.toPatchableAndMergeable(root)

  const finishAddIns: Sheeter.FinishAddIns = {
    $mediaq: Sheeter.mediaqFinishAddInCreator()
  }
  const addInRulesetFilters: Sheeter.RulesetPatchGetters = {
    $mediaq: Sheeter.mediaqRulesetPatchGetterCreator(200), // screen width = 200px
    $whenUsed: Sheeter.whenUsedRulesetFilter
  }
  window.isWeb = true
  const mergedWeb = Sheeter.mergeSheetsAndFinish(patchable, [], finishAddIns)
  const codeRootWeb = Sheeter.mergeRulesetsForCode(mergedWeb, addInRulesetFilters, [mergedWeb.root, false as any, mergedWeb.b, null as any])

  window.isWeb = false
  const mergedNative = Sheeter.mergeSheetsAndFinish(patchable, [], finishAddIns)
  const codeRootNative = Sheeter.mergeRulesetsForCode(mergedNative, addInRulesetFilters, [mergedNative.root, false as any, mergedNative.b, null as any])

  debugger
}

//****************************
// INPUT DATA

const root: Sheeter.Sheet = {
  root: {
    a0: 1,
    $web: {
      a1: 1,
      ":active": {
        a2: 1,
        $before: {
          $mediaq: {
            '100-400': {
              a3: 1,
              $whenUsed: {
                b: {
                  a4: 1,
                  $after: {
                    $web: {
                      a5: 1,
                      ":hover": {
                        a6: 1,
                        $mediaq: {
                          '200-300': {
                            a7: 1,
                          }
                        } as Sheeter.Sheet
                      },
                    },
                    ":hover": {
                      $mediaq: {
                        '300-400': {
                          a8: 1,
                        }
                      } as Sheeter.Sheet
                    }
                  },
                },
              } as Sheeter.Sheet,
            },
          } as Sheeter.Sheet,
        },
      },
    },
  },
  b: {}
}

//window['isWeb'] = true

//****************************
// 

const _toPatchableAndMergeable = {
  "root": {
    "a0": 1,
    "a1": 1,
    ":active": {
      "a2": 1
    }
  },
  "b": {},
  "$addInsFinal": {
    "$mediaq": {
    }
  },
  "$addIns": {
    "$mediaq": {
      "$addIns/$whenUsed/$addIns/$mediaq/root/:active/100-200/b/:hover": {
        "300-400": {
          "a8": 1
        },
        "200-300": {
          "a7": 1
        },
        "#path": [
          "$addIns",
          "$whenUsed",
          "$addIns/$mediaq/root/:active/100-200",
          "b",
          ":hover"
        ],
      },
      "root/:active": {
        "100-200": {
          "a3": 1
        },
        "#path": [
          "root",
          ":active"
        ]
      }
    },
    "$whenUsed": {
      "$addIns/$mediaq/root/:active/100-200": {
        "b": {
          "a4": 1,
          ":hover": {
            "a6": 1
          },
          "a5": 1
        },
        "#path": [
          "$addIns",
          "$mediaq",
          "root/:active",
          "100-200"
        ]
      }
    }
  }
}

const inGetPatchLow = {
  "rootPatches": [
    {
      "path": [
        "root",
        ":active"
      ],
      "items": [
        {
          "a3": 1,
          "a4": 1,
          ":hover": {
            "a6": 1,
            "a8": 1,
            "a7": 1
          },
          "a5": 1
        }
      ]
    }
  ],
  "addInPatches": [
    {
      "path": [
        "$addIns",
        "$whenUsed",
        "$addIns/$mediaq/root/:active/100-200",
        "b",
        ":hover"
      ],
      "items": [
        {
          "a8": 1
        },
        {
          "a7": 1
        }
      ]
    },
    {
      "path": [
        "$addIns",
        "$mediaq",
        "root/:active",
        "100-200"
      ],
      "items": [
        {
          "a4": 1,
          ":hover": {
            "a6": 1,
            "a8": 1,
            "a7": 1
          },
          "a5": 1
        }
      ]
    }
  ]
}

const result = {
  "a0": 1,
  "a1": 1,
  ":active": {
    "a2": 1,
    "a3": 1,
    "a4": 1,
    ":hover": {
      "a6": 1,
      "a8": 1,
      "a7": 1
    },
    "a5": 1
  }
}