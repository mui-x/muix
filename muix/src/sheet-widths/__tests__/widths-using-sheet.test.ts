import { atomizeSheet } from "reactxx-sheeter";
import { initPlatform, toClassNames, Shape } from "reactxx-tests";
import { createSheet, query } from "./widths-sheet.test";

/*
describe("SWITCH using sheet", () => {
  it("NATIVE", () => {
    initPlatform(false);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames('NATIVE, OPENED', query(true), [
      sheet.root,
      sheet.nativeOnly
    ], res => res.toMatchInlineSnapshot())

    toClassNames('NATIVE, CLOSED', query(false), [
      sheet.root,
      sheet.nativeOnly
    ], res => res.toMatchInlineSnapshot())
  })
  it("WEB", () => {
    initPlatform(true);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames('WEB, OPENED', query(true), [
      sheet.root,
      sheet.webOnly
    ], res => res.toMatchInlineSnapshot())

    toClassNames('WEB, CLOSED', query(false), [
      sheet.root,
      sheet.webOnly
    ], res => res.toMatchInlineSnapshot())
  });

});
*/

describe("SWITCH using sheet", () => {
  it("NATIVE", () => {
    initPlatform(false);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames(
      "NATIVE, OPENED",
      query(true),
      [sheet.root, sheet.nativeOnly],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  NATIVE, OPENED
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "backgroundColor: blue /*root/$sheetSwitch.mobileWidth*/",
    "backgroundColor: cyan /*root/$sheetSwitch.tabletWidth*/",
    "backgroundColor: green /*root/$sheetSwitch.desktopWidth*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "backgroundColor": Object {
      "tracePath": "root/$sheetSwitch.desktopWidth",
      "value": "green",
    },
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "backgroundColor": "green",
  },
]
`)
    );

    toClassNames(
      "NATIVE, CLOSED",
      query(false),
      [sheet.root, sheet.nativeOnly],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  NATIVE, CLOSED
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "backgroundColor: blue /*root/$sheetSwitch.mobileWidth*/",
    "backgroundColor: cyan /*root/$sheetSwitch.tabletWidth*/",
    "backgroundColor: green /*root/$sheetSwitch.desktopWidth*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Object {
    "backgroundColor": Object {
      "tracePath": "root/$sheetSwitch.desktopWidth",
      "value": "green",
    },
  },
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "backgroundColor": "green",
  },
]
`)
    );
  });
  it("WEB", () => {
    initPlatform(true);

    const sheet = atomizeSheet<Shape>(createSheet());

    toClassNames("WEB, OPENED", query(true), [sheet.root, sheet.webOnly], res =>
      res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  WEB, OPENED
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".a { background-color:blue /*root/$sheetSwitch.mobileWidth*/ }",
    ".b { background-color:cyan /*root/$sheetSwitch.tabletWidth*/ }",
    ".c { background-color:green /*root/$sheetSwitch.desktopWidth*/ }",
    undefined,
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.mobileWidth*/ }",
    ".e:hover { background-color:lightcyan /*root/$web/:hover/$sheetSwitch.tabletWidth*/ }",
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.desktopWidth*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.desktopWidth*/ }",
    ".c { background-color:green /*root/$sheetSwitch.desktopWidth*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "d c",
]
`)
    );

    toClassNames(
      "WEB, CLOSED",
      query(false),
      [sheet.root, sheet.webOnly],
      res =>
        res.toMatchInlineSnapshot(`
Array [
  "
##########################################
##########################################
#
#  WEB, CLOSED
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    ".a { background-color:blue /*root/$sheetSwitch.mobileWidth*/ }",
    ".b { background-color:cyan /*root/$sheetSwitch.tabletWidth*/ }",
    ".c { background-color:green /*root/$sheetSwitch.desktopWidth*/ }",
    undefined,
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.mobileWidth*/ }",
    ".e:hover { background-color:lightcyan /*root/$web/:hover/$sheetSwitch.tabletWidth*/ }",
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.desktopWidth*/ }",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  Array [
    ".d:hover { background-color:lightblue /*root/$web/:hover/$sheetSwitch.desktopWidth*/ }",
    ".c { background-color:green /*root/$sheetSwitch.desktopWidth*/ }",
  ],
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "d c",
]
`)
    );
  });
});
