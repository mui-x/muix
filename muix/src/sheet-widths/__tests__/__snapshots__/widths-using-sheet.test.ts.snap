// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SWITCH using sheet NATIVE 1`] = `
Array [
  "
##########################################
##########################################
#
#  NATIVE, MOBILE
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 0 /*sheet.root/$widths.0-640*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  "
margin: 0 /*sheet.root/$widths.0-640*/",
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "margin": 0,
  },
]
`;

exports[`SWITCH using sheet NATIVE 2`] = `
Array [
  "
##########################################
##########################################
#
#  NATIVE, TABLET
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  Array [
    "margin: 10 /*sheet.root/$widths.640-1024*/",
  ],
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  "
margin: 10 /*sheet.root/$widths.640-1024*/",
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  Object {
    "margin": 10,
  },
]
`;

exports[`SWITCH using sheet WEB 1`] = `
Array [
  "
##########################################
##########################################
#
#  WEB, MOBILE
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  "
@media (max-width: 639px).a { margin:0 /*sheet.root/$widths.0-640*/ }
@media (min-width: 640px) and (max-width: 1023px).b { margin:10px /*sheet.root/$widths.640-1024*/ }
@media (min-width: 1024px).c { margin:20px /*sheet.root/$widths.1024-*/ }
@media (max-width: 639px).d:hover { color:blue /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (max-width: 639px).e:hover { padding:0 /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (min-width: 640px) and (max-width: 1023px).f:hover { color:red /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 640px) and (max-width: 1023px).g:hover { padding:10px /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 1024px).h:hover { padding:20px /*sheet.root/$web/:hover/$widths.1024-*/ }",
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  "
@media (min-width: 1024px).h:hover { padding:20px /*sheet.root/$web/:hover/$widths.1024-*/ }
@media (min-width: 640px) and (max-width: 1023px).g:hover { padding:10px /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 640px) and (max-width: 1023px).f:hover { color:red /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (max-width: 639px).e:hover { padding:0 /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (max-width: 639px).d:hover { color:blue /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (min-width: 1024px).c { margin:20px /*sheet.root/$widths.1024-*/ }
@media (min-width: 640px) and (max-width: 1023px).b { margin:10px /*sheet.root/$widths.640-1024*/ }
@media (max-width: 639px).a { margin:0 /*sheet.root/$widths.0-640*/ }",
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "h g f e d c b a",
]
`;

exports[`SWITCH using sheet WEB 2`] = `
Array [
  "
##########################################
##########################################
#
#  WEB, TABLET
#
##########################################
##########################################
",
  "
******************************************
*  QUERY TO STHEET
******************************************
",
  "
@media (max-width: 639px).a { margin:0 /*sheet.root/$widths.0-640*/ }
@media (min-width: 640px) and (max-width: 1023px).b { margin:10px /*sheet.root/$widths.640-1024*/ }
@media (min-width: 1024px).c { margin:20px /*sheet.root/$widths.1024-*/ }
@media (max-width: 639px).d:hover { color:blue /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (max-width: 639px).e:hover { padding:0 /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (min-width: 640px) and (max-width: 1023px).f:hover { color:red /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 640px) and (max-width: 1023px).g:hover { padding:10px /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 1024px).h:hover { padding:20px /*sheet.root/$web/:hover/$widths.1024-*/ }",
  "
******************************************
*  APPLY LASTWINS STRATEGY
******************************************
",
  "
@media (min-width: 1024px).h:hover { padding:20px /*sheet.root/$web/:hover/$widths.1024-*/ }
@media (min-width: 640px) and (max-width: 1023px).g:hover { padding:10px /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (min-width: 640px) and (max-width: 1023px).f:hover { color:red /*sheet.root/$web/:hover/$widths.640-1024*/ }
@media (max-width: 639px).e:hover { padding:0 /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (max-width: 639px).d:hover { color:blue /*sheet.root/$web/:hover/$widths.0-640*/ }
@media (min-width: 1024px).c { margin:20px /*sheet.root/$widths.1024-*/ }
@media (min-width: 640px) and (max-width: 1023px).b { margin:10px /*sheet.root/$widths.640-1024*/ }
@media (max-width: 639px).a { margin:0 /*sheet.root/$widths.0-640*/ }",
  "
******************************************
*  FINALIZE FOR for web's className or native's style
******************************************
",
  "h g f e d c b a",
]
`;

exports[`WIDTHS define sheet NATIVE 1`] = `
Object {
  "root": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          "margin: 0 /*sheet.root/$widths.0-640*/",
        ],
        "conditions": Array [
          Object {
            "interval": Array [
              0,
              640,
            ],
            "type": "$widths",
          },
        ],
      },
      Object {
        "atomicArray": Array [
          "margin: 10 /*sheet.root/$widths.640-1024*/",
        ],
        "conditions": Array [
          Object {
            "interval": Array [
              640,
              1024,
            ],
            "type": "$widths",
          },
        ],
      },
      Object {
        "atomicArray": Array [
          "margin: 20 /*sheet.root/$widths.1024-*/",
        ],
        "conditions": Array [
          Object {
            "interval": Array [
              1024,
              0,
            ],
            "type": "$widths",
          },
        ],
      },
    ],
    "name": "root",
    "~": "c",
  },
}
`;

exports[`WIDTHS define sheet WEB 1`] = `
Object {
  "root": Object {
    "list": Array [
      Object {
        "atomicArray": Array [
          "@media (max-width: 639px).a { margin:0 /*sheet.root/$widths.0-640*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          "@media (min-width: 640px) and (max-width: 1023px).b { margin:10px /*sheet.root/$widths.640-1024*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          "@media (min-width: 1024px).c { margin:20px /*sheet.root/$widths.1024-*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          "@media (max-width: 639px).d:hover { color:blue /*sheet.root/$web/:hover/$widths.0-640*/ }",
          "@media (max-width: 639px).e:hover { padding:0 /*sheet.root/$web/:hover/$widths.0-640*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          "@media (min-width: 640px) and (max-width: 1023px).f:hover { color:red /*sheet.root/$web/:hover/$widths.640-1024*/ }",
          "@media (min-width: 640px) and (max-width: 1023px).g:hover { padding:10px /*sheet.root/$web/:hover/$widths.640-1024*/ }",
        ],
      },
      Object {
        "atomicArray": Array [
          "@media (min-width: 1024px).h:hover { padding:20px /*sheet.root/$web/:hover/$widths.1024-*/ }",
        ],
      },
    ],
    "name": "root",
    "~": "c",
  },
}
`;
